### PR TITLE
bulk move motions in call list

### PR DESCRIPTION
**Pull request by MaximilianKrambach**
An attempt for moving around many call list items at once.

This treats the selected items as a tree; keeping the internal hierarchy of the selected items, and inserts them either as sort_parent child of another item, or rearranges the weight(s).

### CODE REVIEW

**Comment by MaximilianKrambach** [comment]
I think I got a working version now (moving as child of another, moving after another), but it could still need some discussion and simplification ( @FinnStutzenstein I could use your help here, as I'm told you are a friend of trees)

**Comment by MaximilianKrambach** [comment]
@emanuelschuetze  This should work now; but we may have different views on corner cases etc.; so a functionality test would be good

**Comment by emanuelschuetze** [comment]
'insert after' action does not work (no request)

I think we need a bulk mode view on server (one request only)...

File: client/src/app/core/ui-services/tree.service.ts [hunk 1]
```diff
@@ -139,4 +139,60 @@ export class TreeService extends OpenSlidesComponent {
...
+     * @returns the selection of items that belong to different branches
+     */
+    public getTopItemsFromTree<T extends Identifiable & Displayable>(tree: OSTreeNode<T>[], items: T[]): T[] {
+        let results: T[] = [];
+        tree.forEach(branch => {
+            const i = this.getTopItemsFromBranch(branch, items);
+            if (i.length) {
+                results = results.concat(i);
+            }
+        });
```
Comment (tsiegleauq) [inline client/src/app/core/ui-services/tree.service.ts]: try
```ts
return tree.map(branch => {
    const top = this.getTopItemsFromBranch(branch, items);
    if (top.length) {
        return top;
    }
}) as T[];
```
(not tested)

File: client/src/app/core/ui-services/tree.service.ts [hunk 2]
```diff
@@ -139,4 +139,60 @@ export class TreeService extends OpenSlidesComponent {
...
+        return results;
+    }
+
+    /**
+     * Return all items not being hierarchically dependant on the items in the input arrray
+     * @param tree
```
Comment (tsiegleauq) [inline client/src/app/core/ui-services/tree.service.ts]: line spacing

File: client/src/app/site/motions/services/motion-multiselect.service.ts [hunk 3]
```diff
@@ -247,4 +250,35 @@ export class MotionMultiselectService {
...
+        const title = this.translate.instant('This will assign the selected motions as belonging to:');
+        const options = ['Set as parent', 'Insert after'];
+        const allMotions = this.repo.getViewModelList();
+        const tree = this.treeService.makeTree(allMotions, 'weight', 'sort_parent_id');
+        const itemsToMove = this.treeService.getTopItemsFromTree(tree, motions);
+        const selectableItems = this.treeService.getTreeWithoutSelection(tree, motions);
+
+        const selectedChoice = await this.choiceService.open(title, selectableItems, false, options);
+        if (selectedChoice) {
+            if (selectedChoice.action === options[0]) {
```
Comment (tsiegleauq) [inline client/src/app/site/motions/services/motion-multiselect.service.ts]: sounds like options[0] and options[1] could reference an enum.
Would improve readability to access `options[actions.whatever-zero-means-here]`

File: client/src/app/core/ui-services/tree.service.ts (hunk 1 above)
Comment (MaximilianKrambach) [reply client/src/app/core/ui-services/tree.service.ts]: So, I tried copypasting your approach, and it doesn't fit. Am I to debug your approach or am I allowed to keep my oh so ugly, but working one?

File: client/src/app/site/motions/services/motion-multiselect.service.ts (hunk 3 above)
Comment (MaximilianKrambach) [reply client/src/app/site/motions/services/motion-multiselect.service.ts]: "options" is only used inside this function, and declared ten lines earlier. I could check for the string itself (which would make it more readable, but may introduce errors if the choices are to be changed, if that helps readability. (If this is desired, I should change similar cases at user-list.component, too)

File: client/src/app/core/ui-services/tree.service.ts (hunk 1 above)
Comment (tsiegleauq) [reply client/src/app/core/ui-services/tree.service.ts]: Okay. There would be `.flat` and `.flatMap` for exactly that usecase, but our setup does not yet support that.

File: client/src/app/site/motions/services/motion-multiselect.service.ts (hunk 3 above)
Comment (tsiegleauq) [reply client/src/app/site/motions/services/motion-multiselect.service.ts]: OK. In case the functionality grows, we might want to put options in a struct/enum.
A comment above the declaration of the options array will be sufficient for now.
